src/main/kotlin/uk/gov/dluhc/printapi/config/ElectorDocumentPdfTemplateProperties.kt

```diff
@@ -2,22 +2,40 @@ package uk.gov.dluhc.printapi.config
 
 import org.springframework.boot.context.properties.ConfigurationProperties
 import org.springframework.boot.context.properties.ConstructorBinding
+import uk.gov.dluhc.printapi.config.ElectorDocumentPdfTemplateProperties.AnonymousElectorDocumentPlaceholder
+import uk.gov.dluhc.printapi.config.ElectorDocumentPdfTemplateProperties.English
+import uk.gov.dluhc.printapi.config.ElectorDocumentPdfTemplateProperties.Welsh
 
 @ConfigurationProperties(prefix = "temporary-certificate.certificate-pdf")
 @ConstructorBinding
-class TemporaryCertificatePdfTemplateProperties(english: English, welsh: Welsh) :
-    ElectorDocumentPdfTemplateProperties(english, welsh)
+class TemporaryCertificatePdfTemplateProperties(
+    english: English<English.TemporaryCertificatePlaceholder>,
+    welsh: Welsh<Welsh.TemporaryCertificatePlaceholder>
+) : ElectorDocumentPdfTemplateProperties<English.TemporaryCertificatePlaceholder, Welsh.TemporaryCertificatePlaceholder>(
+    english,
+    welsh
+)
 
-abstract class ElectorDocumentPdfTemplateProperties(
-    val english: English,
-    val welsh: Welsh
+@ConfigurationProperties(prefix = "anonymous-elector-document.certificate-pdf")
+@ConstructorBinding
+class AnonymousElectorDocumentPdfTemplateProperties(
+    english: English<AnonymousElectorDocumentPlaceholder>,
+    welsh: Welsh<AnonymousElectorDocumentPlaceholder>
+) : ElectorDocumentPdfTemplateProperties<AnonymousElectorDocumentPlaceholder, AnonymousElectorDocumentPlaceholder>(
+    english,
+    welsh
+)
+
+abstract class ElectorDocumentPdfTemplateProperties<EP, WP>(
+    val english: English<EP>,
+    val welsh: Welsh<WP>
 ) {
-    data class English(
+    data class English<T>(
         val path: String,
-        val placeholder: Placeholder,
+        val placeholder: T,
         val images: Images
     ) {
-        data class Placeholder(
+        data class TemporaryCertificatePlaceholder(
```

Comment:
Hi @Neil-Massey ,

I need to admit, it took some time for me to understand the new structure, especially because we have `TemporaryCertificatePlaceholder` classes under `English` and `Welsh` classes but not the `AnonymousElectorDocumentPlaceholder`. The reason for it is, I was expecting AED structure would be similar to Temporary Certificate. I can now see that the AED placeholders don't change by the language and that is why it is not duplicated under English and Welsh classes. 

However, I assume it would be much easier and quicker for me to understand if we have it as below:

```
@ConfigurationProperties(prefix = "temporary-certificate.certificate-pdf")
@ConstructorBinding
class TemporaryCertificatePdfTemplateProperties(
    english: English<EnglishTemporaryCertificatePlaceholder>,
    welsh: Welsh<WelshTemporaryCertificatePlaceholder>
) : ElectorDocumentPdfTemplateProperties<EnglishTemporaryCertificatePlaceholder, WelshTemporaryCertificatePlaceholder>(
    english,
    welsh
)

@ConfigurationProperties(prefix = "anonymous-elector-document.certificate-pdf")
@ConstructorBinding
class AnonymousElectorDocumentPdfTemplateProperties(
    english: English<AnonymousElectorDocumentPlaceholder>,
    welsh: Welsh<AnonymousElectorDocumentPlaceholder>
) : ElectorDocumentPdfTemplateProperties<AnonymousElectorDocumentPlaceholder, AnonymousElectorDocumentPlaceholder>(
    english,
    welsh
)

abstract class ElectorDocumentPdfTemplateProperties<EP, WP>(
    val english: English<EP>,
    val welsh: Welsh<WP>
) {
    data class English<T>(
        val path: String,
        val placeholder: T,
        val images: Images
    )

    data class Welsh<T>(
        val path: String,
        val placeholder: T,
        val images: Images
    )

    data class EnglishTemporaryCertificatePlaceholder(
        val electorName: String,
        val localAuthorityNameEn: String,
        val dateOfIssue: String,
        val validOnDate: String,
        val certificateNumber: String
    )

    data class WelshTemporaryCertificatePlaceholder(
        val electorName: String,
        val localAuthorityNameEn: String,
        val localAuthorityNameCy: String,
        val dateOfIssue: String,
        val validOnDate: String,
        val certificateNumber: String
    )

    data class AnonymousElectorDocumentPlaceholder(
        val electoralRollNumber: String,
        val dateOfIssue: String,
        val certificateNumber: String
    )

    data class Images(
        val voterPhoto: PhotoProperties
    )
    
    data class PhotoProperties(
        val pageNumber: Int = 1,
        val absoluteXMm: Float,
        val absoluteYMm: Float,
        val fitWidthMm: Float,
        val fitHeightMm: Float
    )
}

```
We wouldn't be duplicating the image class as well in this approach.
What are your thoughts?

Reply:
Looks good - I wanted to revisit after I had a working version committed (I tried several approaches before this generified ones). Will update with your suggestion 👍